torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,404 @@
+            mlir::TypeID::get<BuildStableHLOCompositePass>()) {}
+
+  ~BuildStableHLOCompositePass() override = default;
+
+  void runOnOperation() override {
+    mlir::ModuleOp module_op = getOperation();
+    llvm::SmallVector<mlir::func::FuncOp> func_ops(
+        module_op.getOps<mlir::func::FuncOp>());
+    for (mlir::func::FuncOp& func_op : func_ops) {
+      llvm::DenseMap<const mlir::Operation*, size_t> op_line_num =
+          BuildOperationsLineNumberMap(func_op);
+      for (auto op : func_op.getOps<mlir::stablehlo::CustomCallOp>()) {
+        BuildStableHLOCompositeOp(op.getOperation(), op_line_num);
+      }
+    }
+  }
+
+  mlir::StringRef getName() const override {
+    return llvm::getTypeName<BuildStableHLOCompositePass>();
+  }
+
+  std::unique_ptr<mlir::Pass> clonePass() const override {
+    return std::make_unique<BuildStableHLOCompositePass>(*this);
+  }
+
+ private:
+  llvm::DenseMap<const mlir::Operation*, size_t> BuildOperationsLineNumberMap(
+      mlir::func::FuncOp func_op) const {
+    llvm::DenseMap<const mlir::Operation*, size_t> op_line_num;
+    for (const auto& op : llvm::enumerate(func_op.getOps())) {
+      op_line_num[&op.value()] = op.index();
+    }
+    return op_line_num;
+  }
+
+  std::unique_ptr<BoundaryMetadata> GetBoundaryMetadata(mlir::Operation* op) {
+    if (!IsXlaMarkTensorOp(op)) {
+      return nullptr;
+    }
+    auto backend_config =
+        op->getAttr("backend_config").dyn_cast<mlir::StringAttr>();
+    if (backend_config == nullptr) {
+      return nullptr;
+    }
+    return BoundaryMetadata::Parse(backend_config);
+  }
+
+  mlir::DictionaryAttr BuildDictionaryAttrFromJsonMap(
+      mlir::OpBuilder& builder,
+      const std::unordered_map<std::string, json>& json_map) {
+    llvm::SmallVector<mlir::NamedAttribute> named_attrs;
+    for (auto& [key, j] : json_map) {
+      switch (j.type()) {
+        case json::value_t::number_integer:
+        case json::value_t::number_unsigned:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getI64IntegerAttr(j.template get<int64_t>())});
+          break;
+        case json::value_t::number_float:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getF32FloatAttr(j.template get<float>())});
+          break;
+        case json::value_t::boolean:
+          named_attrs.push_back({builder.getStringAttr(key),
+                                 builder.getBoolAttr(j.template get<bool>())});
+          break;
+        case json::value_t::string:
+          named_attrs.push_back(
+              {builder.getStringAttr(key),
+               builder.getStringAttr(j.template get<std::string>())});
+          break;
+        default:
+          // Ignored unrecognizable attr json
+          break;
+      }
+    }
+    return builder.getDictionaryAttr(named_attrs);
+  }
+
+  void BuildStableHLOCompositeOp(
```

Comment:
The cyclomatic complexity of this function is pretty high, for loops nested in while loops, nested ifs in loops etc. Would be good to break down into smaller functions. Would help with understanding what exactly is happening as well.